lib/collection/src/recommendations.rs

```diff
@@ -0,0 +1,309 @@
+    Fut: Future<Output = Option<RwLockReadGuard<'a, Collection>>>,
+{
+    // shortcuts batch if all requests with limit=0
+    if request_batch.searches.iter().all(|s| s.limit == 0) {
+        return Ok(vec![]);
+    }
+    // pack all reference vector ids
+    let mut all_reference_vectors_ids: HashMap<_, HashSet<PointIdType>> = Default::default();
+    let mut vector_names_per_collection: HashMap<_, HashSet<String>> = Default::default();
+
+    for request in &request_batch.searches {
+        if request.positive.is_empty() {
+            return Err(CollectionError::BadRequest {
+                description: "At least one positive vector ID required".to_owned(),
+            });
+        }
+        let collection_name = request.lookup_from.as_ref().map(|x| &x.collection);
+
+        let reference_vectors_ids = all_reference_vectors_ids
+            .entry(collection_name)
+            .or_insert_with(HashSet::new);
+
+        let vector_names = vector_names_per_collection
+            .entry(collection_name)
+            .or_insert_with(HashSet::new);
+
+        vector_names.insert(get_search_vector_name(request));
+
+        for point_id in request.positive.iter().chain(&request.negative) {
+            reference_vectors_ids.insert(*point_id);
+        }
+    }
+
+    debug_assert!(all_reference_vectors_ids.len() == vector_names_per_collection.len());
+
+    let mut collections_names: Vec<_> = Default::default();
+    let mut vector_retrieves: Vec<_> = Default::default();
+    for (collection_name, reference_vectors_ids) in all_reference_vectors_ids.into_iter() {
+        collections_names.push(collection_name);
+        let points: Vec<_> = reference_vectors_ids.into_iter().collect();
+        let vector_names: Vec<_> = vector_names_per_collection
+            .remove(&collection_name)
+            .unwrap()
+            .into_iter()
+            .collect();
+        match collection_name {
+            None => vector_retrieves.push(retrieve_points_with_locked_collection(
+                CollectionRefHolder::Ref(collection),
+                points,
+                vector_names,
+            )),
+            Some(name) => {
+                let other_collection = collection_by_name(name.to_string()).await;
+                match other_collection {
+                    Some(other_collection) => {
+                        vector_retrieves.push(retrieve_points_with_locked_collection(
+                            CollectionRefHolder::Guard(other_collection),
+                            points,
+                            vector_names,
+                        ))
+                    }
+                    None => {
+                        return Err(CollectionError::NotFound {
+                            what: format!("Collection {}", name),
+                        })
+                    }
+                }
+            }
+        }
+    }
+
+    let all_reference_vectors: Vec<Vec<Record>> = try_join_all(vector_retrieves).await?;
```

Comment:
should we validate that the vectors have a size equal to the current collection vector size?

Reply:
I expected it to fail on it's own, will check